Sources/Workspace/DefaultPluginScriptRunner.swift

```diff
@@ -125,204 +127,257 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner, Cancellable {
-                compilerInputsHash = .none
+                // We couldn't read the compilation state file even though it existed. We warn about it but proceed with recompiling.
+                observabilityScope.emit(debug: "... couldn't read previous compilation state (\(error))")
             }
+        }
+        
+        // If we still have a compilation state, it means the executable is still valid and we don't need to do anything.
+        if let compilationState = compilationState {
+            // Just call the completion handler with the persisted results.
+            let result = PluginCompilationResult(
+                succeeded: compilationState.succeeded,
+                commandLine: commandLine,
+                executableFile: execFilePath,
+                diagnosticsFile: diagFilePath,
+                compilerOutput: compilationState.output,
+                cached: true)
+            return callbackQueue.async {
+                completion(.success(result))
+            }
+        }
 
-            // If we already have a compiled executable, then compare its hash with the new one.
-            var compilationNeeded = true
-            let hashFile = executableFile.parentDirectory.appending(component: execName + ".inputhash")
-            if fileSystem.exists(executableFile) && fileSystem.exists(hashFile) {
+        // Otherwise we need to recompile. We start by cleaning up any old files to avoid confusion if the compiler can't be invoked.
+        do {
+            try fileSystem.removeFileTree(execFilePath)
+            try fileSystem.removeFileTree(diagFilePath)
+            try fileSystem.removeFileTree(stateFilePath)
+        }
+        catch {
+            observabilityScope.emit(debug: "... couldn't clean up before invoking compiler (\(error))")
+        }
+        
+        // Now invoke the compiler asynchronously.
+        let startTime = DispatchTime.now()
+        Process.popen(arguments: commandLine, environment: toolchain.swiftCompilerEnvironment, queue: callbackQueue) {
+            // We are now on our caller's requested callback queue, so we just call the completion handler directly.
+            dispatchPrecondition(condition: .onQueue(callbackQueue))
+            completion($0.tryMap { process in
+                // Emit the compiler output as observable info.
+                let compilerOutput = ((try? process.utf8Output()) ?? "") + ((try? process.utf8stderrOutput()) ?? "")
+                observabilityScope.emit(info: compilerOutput)
+
+                // Save the persisted compilation state for possible reuse next time.
+                let compilationState = PersistedCompilationState(
+                    commandLine: commandLine,
+                    environment: toolchain.swiftCompilerEnvironment,
+                    inputHash: compilerInputHash,
+                    duration: Double(startTime.distance(to: .now()).milliseconds() ?? 0) / 1000.0,
```

Comment:
Millisecond precision seems enough here.

Reply:
Good point, though because the idea is to encode this into JSON, that would lead to additional code to encode or decode it.

I looked at adding a `fileprivate` extension to `DispatchTimeInterval` to make it adopt `Codable`, but that's not permissible — we then have to make it codable for the whole module.

Since we're currently not using the duration maybe I will just take it out for now.

Reply:
one way to do this is use a property wrapper which implements the serialization for the duration property. this is how we often deal with date formatting

Reply:
I see.  As this is unused, I'll go ahead and remove it, but I'll keep that in mind for next time.